docs/DesignDocs/DebuggerSupportForMultiModuleEnclaves.md

```diff
@@ -0,0 +1,268 @@
+		uint64_t base_address;
+		uint64_t size;
+	} oe_debug_module_t;
+	
+```
+
+The `path` and `path_length` together specify the path of the module's ELF
+binary. The `base_address` and `size` together specify the memory region
+where the enclave has been loaded.
+
+The following notification functions are used to notify the debugger about the
+loading and unloading of a module.
+
+- `oe_result_t oe_result_t oe_debug_notify_module_loaded(oe_debug_module_t* module)`
+
+   This function is called by the SDK's host-runtime after a module has been
+   loaded, but before it has been initialized. A module is initialized by invoking
+   it's constructor functions. It doesn not matter whether the module has been relocated
+   or not before the debugger is notified about the module.
+
+   In situations where a module may be dynamically loaded by an enclave, as is the case
+   when SGX LKL/openlibos launch applications within the enclave, this notification
+   function is called by the SDK's enclave-side runtime prior to the initialization
+   of the module.
+
+- `oe_result_t oe_debug_notify_module_unloaded(oe_debug_module_t* module)`
+
+   This function is called by the SDK runtimes (host-side or enclave-side), after a
+   module has been unloaded.
+
+Debugger Handling of Module Load/Unload Notifications
+
+- Linux
+  Similar to enclave load/unload, the debugger sets internal breakpoints in
+  `oe_debug_notify_module_loaded` and `oe_debug_notify_module_unloaded`
+  functions. When the SDK runtimes call these functions, the breakpoint is
+  triggered and the debugger get control. The debugger reads the module information
+  from the `oe_debug_module_t` structure either loads or unloads the debugging
+  information for the module.
+
+- Windows
+  - Host Side
+  
+    The SDK host-side runtime uses the `RaiseException` Windows API to
+	communicate with the debugger.
+	`oe_debug_notify_module_loaded` raises the following exception:
+	```c
+ 
+        ULONG_PTR args[1] = {(ULONG_PTR)module};
+		RaiseException(
+		  OE_DEBUGRT_MODULE_LOADED_EVENT,
+		  0, // dwFlags
+		  1, // number of args
+		  args);
+		  
+    ```
+	
+     whereas `oe_debug_notify_enclave_terminated` raises:
+     ```c
+   
+        ULONG_PTR args[1] = {(ULONG_PTR)module};
+        RaiseException(
+          OE_DEBUGRT_MODULE_UNLOADED_EVENT,
+          0, // dwFlags
+          1, // number of args
+          args);
+		  
+	```
+	
+  - Enclave side
+    
+	The enclave side runtime cannot call Windows API functions. Instead, it
+	relies on `int 3` exception to communicate with the debugger. The `int 3`
+	instruction is preceeded by special byte sequence `0x0edeb0gca11bac00` to
+	allow the debugger to distinguish between these special `int 3` exceptions
+	and any other `int 3` exceptions that may occur.
+	
+	`oe_debug_notify_module_loaded` uses the following instruction pattern:
+	```asm
+
+       mov module, %rax
+	   mov $1, %rbx     // 1 => ModuleLoaded
+	   jmp do_int3
```

Comment:
From the SIG-arch meeting: @binxing had a suggestion about using `nop` and having the debugger overwrite it with the `int 3` as needed, which would make the code path uniform on the enclave side regardless of whether the enclave is under debug. There was an open question as to whether the EDBGWR instruction would fault if the page itself is RX-only.

Reply:
EDBGWR ignores the EPCM RWX permissions (it does require that the enclave is loaded in debug mode).  Thus, the OS interfaces (WriteProcessMemory for Windows and ptrace for Linux) should work the same for when a debugger writes to non-enclave memory.

Reply:
Austen and I discussed this. While this is an elegant proposal, it seems like it is quite close to what the debugger does underneath the hood for a normal breakpoint. Is there any benefit this approach would have in comparison to a hidden breakpoint?